### PR TITLE
feat: fine tune qwen3 for knowledge specialization

### LINKED ISSUE #239: Create Specialized LLM Models for the Lightweight LLM Server for Integration Test
**Is your feature request related to a problem? Please describe.**
Since the router is to support Mixture-of-Models, we need to have integration test suites to verify such functionalities.

Now with #228, we are able to use a real yet lightweight models as LLM server, we can extend such server to support multiple and specialized models.

**Describe the solution you'd like**
- Fine tune the lightweight models, Qwen3 0.6B or Gemma 270m, for different specializations.
- Run the entire model_eval and config generation workflow
- Run benchmark to evaluate the router's performance and accuracy

@Xunzhuo @yossiovadia @yuluo-yx @JaredforReal WDYT?

### PR DESCRIPTION
**Pull request by rootfs**
**What type of PR is this?**

Fine tune qwen3 0.6B for specialization. There are two ways: with MMLU as training dataset (i.e. leakage) for demo purpose, and use other dataset for training (no leakage).

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #239 

Release Notes: Yes/No

### CODE REVIEW

**Comment by rootfs** [comment]
With the no_leakage training, the math accuracy is improved after fine tuning:

```bash 
train_all_specialists_no_leakage.sh 2 100 5
```

```
2025-10-15 20:10:22 - common_lora_utils - INFO - IMPROVEMENT ANALYSIS (No Data Leakage)
2025-10-15 20:10:22 - common_lora_utils - INFO - 📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊📊
2025-10-15 20:10:22 - common_lora_utils - INFO - 
================================================================================
2025-10-15 20:10:22 - common_lora_utils - INFO - OVERALL RESULTS:
2025-10-15 20:10:22 - common_lora_utils - INFO - ================================================================================
2025-10-15 20:10:22 - common_lora_utils - INFO -   Baseline (Untrained):     10.00%
2025-10-15 20:10:22 - common_lora_utils - INFO -   Post-training:            24.00%
2025-10-15 20:10:22 - common_lora_utils - INFO -   Absolute Improvement:     +14.00%
2025-10-15 20:10:22 - common_lora_utils - INFO -   Relative Improvement:     +140.0%
```

File: src/training/training_lora/mmlu_pro_solver_lora/train_all_specialists_no_leakage.sh [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+EPOCHS=${3:-5}
+BATCH_SIZE=2
+LORA_RANK=32
+
+# Output directory
+OUTPUT_BASE_DIR="models_no_leakage"
+LOG_DIR="training_logs_no_leakage"
+
+# Virtual environment
+VENV_PATH="/home/ubuntu/rootfs/semantic-router.bak/.venv"
```
Comment (Xunzhuo) [inline src/training/training_lora/mmlu_pro_solver_lora/train_all_specialists_no_leakage.sh]: this should be removed

File: src/training/training_lora/mmlu_pro_solver_lora/train_all_specialists_no_leakage.sh [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+
+# ============================================================================
+# CONFIGURATION
+# ============================================================================
+
+# Default parameters (can be overridden by command line arguments)
+GPU_ID=${1:-2}
+SAMPLES_PER_DATASET=${2:-1000}
+EPOCHS=${3:-5}
+BATCH_SIZE=2
```
Comment (Xunzhuo) [inline src/training/training_lora/mmlu_pro_solver_lora/train_all_specialists_no_leakage.sh:26]: batch size should be configurable

**Comment by rootfs** [comment]
Latest with open math reasoning dataset:

# Training

```bash
accelerate launch --num_processes=4 --multi_gpu ft_qwen3_mmlu_solver_lora_no_leakage.py     --mode train     --model-type math-reasoner     --epochs 3     --max-samples-per-dataset 1500
```

# Test
```bash
python ft_qwen3_mmlu_solver_lora_no_leakage.py     --mode test     --model-type math-reasoner   --max-tokens-test 5000
```

**Comment by rootfs** [comment]
Trained in Qwen3 4B using command

```bash
$ accelerate launch --num_processes=4 --multi_gpu ft_qwen3_mmlu_solver_lora_no_leakage.py     --mode train     --model-type math-reasoner     --epochs 3 --max-samples-per-dataset 500 --max-tokens-test 5192  --use-cot --filter-category math --model Qwen/Qwen3-4B
```

Test command
```bash
python ft_qwen3_mmlu_solver_lora_no_leakage.py     --mode test     --model-type math-reasoner     --gpu-id 0 --max-tokens-test 5192  --no-cot --filter-category math  --model Qwen/Qwen3-4B
```

Test result:
```
================================================================================
2025-10-18 14:57:11 - common_lora_utils - INFO - 📊 EVALUATION RESULTS COMPARISON
2025-10-18 14:57:11 - common_lora_utils - INFO - ================================================================================
2025-10-18 14:57:11 - common_lora_utils - INFO - Baseline (Untrained): 80.0%
2025-10-18 14:57:11 - common_lora_utils - INFO - Trained Model:        83.3%
2025-10-18 14:57:11 - common_lora_utils - INFO - Improvement:          +3.3%
2025-10-18 14:57:11 - common_lora_utils - INFO - ================================================================================
```

The fine tuned model is pushed 
https://huggingface.co/llm-semantic-router/qwen3_mmlu_math-reasoner_no_leakage_r32
